Add notebooks describing pulses and beamline components

nvaytet opened the pull request:
- Notebook on pulses
- Notebook on components
- Refactored `Pulse` to use `classmethods` to set up pulses differently
- Convert component input values to float to avoid issues such as conversion to radians doing strange things

SimonHeybrock commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+    "    open=sc.array(\n",
+    "        dims=['cutout'],\n",
+    "        values=[60.0, 150.0],\n",
+    "        unit='deg',\n",
```
Didn't you define `deg` above?

nvaytet replied:
Yes but primarily for when multiplying it with scalars. When I need an array, I think I still need to create it like this.

nvaytet replied:
Ah you mean doing `sc.array(..., unit=deg)` instead of `'deg'`? I'm not sure it's crucial?

SimonHeybrock replied:
Not crucial sure, but the other day you said it was confusing to have both strings and named variables when using units. Personally I don't mind.

SimonHeybrock commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "chopper1.tofs.plot()"
```
So the results are stored in the choppers/detectors? I suppose this will go wrong when the same components are used twice, either in the same or in different models?

nvaytet replied:
I liked the fact that you could just do `chopper1.tofs.plot()`, but at the same time it's quite weird to have a `Model` set some data properties onto the components.
Having everything in the `Model` would probably also make things like out-of-date data when choppers have been modified easier (your comment further up).

SimonHeybrock replied:
> So you are saying the model should hold everything?

Not necessarily, one could just change the way components are inserted into the model, for example.

See also comment above, maybe we can sit down and have a think... whether we want to change this or not. I don't think it would be too difficult to achieve.

nvaytet replied:
See change of mechanism update.

SimonHeybrock commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+    "    open=sc.array(\n",
+    "        dims=['cutout'],\n",
+    "        values=[60.0, 100.0, 160.0, 220.0],\n",
+    "        unit='deg',\n",
```
```suggestion
    "        unit=deg,\n",
```

SimonHeybrock commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+   "id": "09d6cbef-f2f3-48db-a696-b4ed912817c8",
+   "metadata": {},
+   "source": [
+    "While the detector now sees 4 peaks"
```
"Peaks" may be confusing language, since it may be associated with Bragg peaks.

nvaytet replied:
Not sure what else to call it, I also have peaks further up. I don't think 'bursts' is any better?

SimonHeybrock replied:
Bunches? Subframes?

nvaytet replied:
I kept peaks for now.

SimonHeybrock commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model.plot(max_rays=100, blocked_rays=5000)"
```
Does this sample from the simulated rays?

SimonHeybrock replied:
Is suppose that is explained in the `plot` docstring?

SimonHeybrock commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+    "## Adding a monitor\n",
+    "\n",
+    "Detectors can be placed anywhere in the beam path,\n",
+    "and in the next example we place it between the first and second choppers."
```
```suggestion
    "and in the next example we place it between the first and second chopper."
```

SimonHeybrock commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+   "source": [
+    "monitor = tof.Detector(distance=15.0 * meter, name='monitor')\n",
+    "\n",
+    "model.detectors = [monitor, detector]\n",
```
Does the order matter?

nvaytet replied:
No, components get sorted according to `.distance`.

SimonHeybrock commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+   "id": "f1d36f43-92d0-4dfd-8c98-6eb3f4a4c599",
+   "metadata": {},
+   "source": [
+    "In the future, we may consider adding monitors with an efficiency profile,\n",
+    "that would have a wavelength-dependent probability of absorbing a neutron."
```
Unless you have very concrete plans to add this, I would not mention it.

SimonHeybrock commented on docs/pulses.ipynb:
```diff
@@ -0,0 +1,251 @@
+    "By default, the time and wavelength distributions of neutrons are both flat (uniform).\n",
+    "We then need to define a pulse duration (start and end times) and a wavelength range.\n",
+    "\n",
+    "To create a pulse with 1M neutrons, uniformly distributed in the ranges of 1-3 ms for birth times,\n",
```
```suggestion
    "To create a pulse with 1 million neutrons, uniformly distributed in the ranges of 1-3 ms for birth times,\n",
```

SimonHeybrock commented on docs/pulses.ipynb:
```diff
@@ -0,0 +1,251 @@
+   "outputs": [],
+   "source": [
+    "pulse = tof.Pulse(\n",
+    "    neutrons=1_000_000, tmin=1.0 * ms, tmax=3.0 * ms, lmin=1.0 * AA, lmax=10.0 * AA\n",
```
What is `l`?

nvaytet replied:
lambda. I'm happy to change. Initially I felt `wavelength_min/wavelength_max` was a bit lengthy, but I am probably changing my mind in the favour of clarity.

SimonHeybrock replied:
`wmin`, `wmax` might be marginally clearer? But I agree that being lengthy is probably better.

SimonHeybrock commented on docs/pulses.ipynb:
```diff
@@ -0,0 +1,251 @@
+   "id": "1bf9e593-fb94-4174-b4f1-7cab27b7cf49",
+   "metadata": {},
+   "source": [
+    "To inspect the data in the pulse, we can either look at it `.birth_times` or `.wavelengths` members"
```
```suggestion
    "To inspect the data in the pulse, we can either look at its `birth_times` or `wavelengths` properties"
```

SimonHeybrock commented on docs/pulses.ipynb:
```diff
@@ -0,0 +1,251 @@
+    "    },\n",
+    ")\n",
+    "\n",
+    "pulse = tof.Pulse.from_distribution(neutrons=200_000, p_time=p_time, p_wav=p_wav)\n",
```
Not sure if you explain it elsewhere, but it is not obvious how the two distributions interact. Are the simply "broadcast" against each other, i.e., this is not a full 2-D definition of an arbitrary pulse?

SimonHeybrock reviewed (commented):
Comments on the notebooks (can look at code later, when some questions are answered, so I have a better picture):

nvaytet commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+    "## Adding a detector\n",
+    "\n",
+    "We first add a `Detector` component which simply records all the neutrons that reach it.\n",
+    "It does not block any neutrons, they all travel through the detector without being absorbed."
```
:+1: see #27

nvaytet commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+    "detector = tof.Detector(distance=30.0 * meter)\n",
+    "\n",
+    "# Build the instrument model\n",
+    "model = tof.Model(pulse=pulse, detectors=[detector])\n",
+    "model"
```
Nothing in particular, I thought it made sense to "run the model". We could change the name, I really don't mind, as long as it's not too long ;-)

nvaytet commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model.choppers = [chopper1]  # Note that model.choppers needs to be a list\n",
```
Yes. Do you mean I should write `needs to be a list or a tuple`?

SimonHeybrock replied:
"an iterable"?

nvaytet commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+   "outputs": [],
+   "source": [
+    "model.choppers = [chopper1]  # Note that model.choppers needs to be a list\n",
+    "model.run()\n",
```
No. You mean like a flag that tells the model it is out of date? It would need to propagate that flag to the components, in case someone inspects the data on the detector after having changed the choppers and not re-run the model.

nvaytet replied:
So I changed the mechanism where calling `model.run()` returns a read-only `Result` object that contains all the data and `.plot()` methods.

nvaytet commented on docs/pulses.ipynb:
```diff
@@ -0,0 +1,251 @@
+   "id": "e735311e-f7a7-4f40-b6b8-73e6388d398e",
+   "metadata": {},
+   "source": [
+    "It is possible to use a finer sampling of the supplied distribution by setting the `sampling` parameter:"
```
> Does it interpolate the given distributions?

Yes, I will make it clearer.

nvaytet commented on docs/components.ipynb:
```diff
@@ -0,0 +1,386 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "detector.tofs.visible.data"
```
Done.

SimonHeybrock commented on docs/pulses.ipynb:
```diff
@@ -53,7 +53,11 @@
     "pulse = tof.Pulse(\n",
-    "    neutrons=1_000_000, tmin=1.0 * ms, tmax=3.0 * ms, lmin=1.0 * AA, lmax=10.0 * AA\n",
+    "    neutrons=1_000_000,\n",
+    "    t_start=1.0 * ms,\n",
+    "    t_end=3.0 * ms,\n",
+    "    wav_min=1.0 * AA,\n",
+    "    wav_max=10.0 * AA,\n",
```
Mixing start/end and min/max seems like something no one will ever be able to remember. Can we use one convention for both `t` and `wav`?

nvaytet replied:
Yeah, I think I will go with min/max :+1:

SimonHeybrock commented on src/tof/result.py:
```diff
@@ -0,0 +1,230 @@
+        self._choppers = choppers
+        self._detectors = detectors
+
+    @property
+    def choppers(self) -> Dict[str, Chopper]:
+        """The choppers in the model."""
+        return self._choppers
```
So users can still change the choppers of the result? Isn't that similar to the problem we had before?

nvaytet replied:
Yes you are right, I know am returning read-only objects in the result.

SimonHeybrock replied:
Can still remove or insert choppers though?

nvaytet replied:
Would you then turn `self._choppers` and `self._detectors` into a read-only dict?

SimonHeybrock replied:
https://docs.python.org/3/library/types.html#types.MappingProxyType, maybe?

SimonHeybrock commented on src/tof/model.py:
```diff
@@ -44,6 +42,24 @@ def __init__(
             self.detectors = [self.detectors]
         self.pulse = pulse
 
+    def add(self, component):
```
```suggestion
    def insert(self, component):
```
?

nvaytet replied:
I think `insert` suggests it is also expecting a positional index where to insert the component. But since we don't care (sorting according to distance from source is performed on the fly), I actually prefer `add`, I think it is more generic.

SimonHeybrock commented on src/tof/model.py:
```diff
@@ -62,177 +78,187 @@ def run(self, npulses: int = 1):
+        result_choppers = {}
+        result_detectors = {}
+        for c in components:
+            comp = Component(**c.to_dict())
```
Took me a while to figure out that you are converting detectors and choppers to components. Some comments on the overall mechanism here (in `run`) would be useful.

SimonHeybrock commented on src/tof/model.py:
```diff
@@ -62,177 +78,187 @@ def run(self, npulses: int = 1):
-            ),
-            axis=1,
+        return Result(
+            pulse=self.pulse, choppers=result_choppers, detectors=result_detectors
```
So these are not instances of `Chopper` and `Detector`?

nvaytet replied:
These are now Readonly chopper-like and detector-like objects

SimonHeybrock commented on src/tof/model.py:
```diff
@@ -62,177 +78,187 @@ def run(self, npulses: int = 1):
-        limited by setting ``max_rays``.
-        In addition, it is possible to also plot the rays that are blocked by
-        choppers along the flight path by setting ``blocked_rays > 0``.
+    # def _add_rays(self, ax, tofs, birth_times, distances, cbar=True, wavelengths=None):
```
commented code

nvaytet replied:
Yes, please note that I am not at all done with my changes, even though I pushed. I sometimes have to do this when I'm in the middle of something but I have to leave. I will send you a message when I am ready for another review.

nvaytet commented:
@SimonHeybrock this is ready for another review, thanks.

SimonHeybrock commented on src/tof/chopper.py:
```diff
@@ -75,3 +79,58 @@ def __repr__(self) -> str:
+        self._close_times = chopper.close_times
+        super().__init__()
+
+    @property
+    def frequency(self) -> sc.Variable:
+        return self._frequency
+
+    @property
+    def open(self) -> sc.Variable:
+        return self._open
+
```
Values of these can of course all be changed, since `sc.Variable` is mutable. But maybe that is fine?

SimonHeybrock commented on src/tof/detector.py:
```diff
@@ -20,9 +22,37 @@ class Detector(Component):
+
+
+@dataclass(frozen=True)
+class ReadonlyDetector(Component):
```
I don't mind the "readonly" much, but I wonder if there is a clearer name for users? Something like `DetectorResult`, or something implying "this is what the detector sees"?

nvaytet replied:
ChatGPT suggests `DetectorReading`, quite good?

SimonHeybrock replied:
:+1:

SimonHeybrock commented on src/tof/model.py:
```diff
@@ -2,16 +2,36 @@
+ComponentType = Union[Chopper, Detector]
+
+
+def _input_to_dict(
+    obj: Optional[
+        Union[
+            Dict[str, ComponentType],
+            List[ComponentType],
+            Tuple[ComponentType, ...],
+            ComponentType,
+        ]
+    ]
+):
```
Would this have to have a default of `None` to be optional? In any case,
```suggestion
def _input_to_dict(
    obj: Union[
            None,
            Dict[str, ComponentType],
            List[ComponentType],
            Tuple[ComponentType, ...],
            ComponentType,
        ]
):
```
would be more readable.

SimonHeybrock commented on src/tof/model.py:
```diff
@@ -31,18 +51,71 @@ class Model:
-        detectors: Union[Detector, List[Detector]],
         pulse: Pulse,
+        choppers: Optional[
+            Union[Chopper, List[Chopper], Tuple[Chopper, ...], Dict[str, Chopper]]
```
Seems odd to accept both `Iterable[Chopper]` (list, tuple) and `Dict`?

nvaytet replied:
It's actually a good point, because if the user provides a dict where the keys do not agree with the `chopper.name`, then it may mess things up elsewhere. So I will remove the option of giving a dict. Thanks

SimonHeybrock commented on src/tof/model.py:
```diff
@@ -31,18 +51,71 @@ class Model:
+    def __getitem__(self, name) -> Union[Chopper, Detector]:
+        if name not in self:
+            raise KeyError(f"No component with name {name} was found.")
+        return self.choppers.get(name, self.detectors.get(name))
```
Note that this gets from `detectors` even if the `choppers.get` succeeds.

nvaytet replied:
:+1:  I changed to
```Py
return self.chopper[name] if name in self.choppers else self.detectors[name]
```

SimonHeybrock commented on src/tof/model.py:
```diff
@@ -31,18 +51,71 @@ class Model:
+                f"Cannot add component of type {type(component)} to the model."
+            )
+
+    def remove(self, name: str):
```
Why? Isn't `__delitem__` enough?

SimonHeybrock commented on src/tof/pulse.py:
```diff
@@ -51,116 +52,226 @@ def __init__(
+        pulse.wmin = pulse.wmin.to(unit='angstrom')
+        pulse.wmax = pulse.wmax.to(unit='angstrom')
+
+        x_time = np.linspace(pulse.tmin.value, pulse.tmax.value, sampling)
+        x_wav = np.linspace(pulse.wmin.value, pulse.wmax.value, sampling)
+        p_time = np.interp(
+            x_time,
+            params.time.coords['time'].to(unit='s').values,
+            params.time.values,
+        )
+        p_time /= p_time.sum()
+        p_wav = np.interp(
+            x_wav,
+            params.wavelength.coords['wavelength'].to(unit='angstrom').values,
+            params.wavelength.values,
+        )
+        p_wav /= p_wav.sum()
```
Have you considered using https://scipp.github.io/generated/modules/scipy/scipp.scipy.interpolate.interp1d.html?highlight=interp1d?

nvaytet replied:
I think originally i went with `numpy` because it was all `numpy` internally, no `scipp`, and I didn't want to add `scipy` as a dependency. But now we have `scipp` and also depend on `scipy` for the constants, so I will try it and see if it makes things cleaner.

SimonHeybrock commented on src/tof/pulse.py:
```diff
@@ -51,116 +52,226 @@ def __init__(
+        p_time = p_time / p_time.sum()
+        p_wav = p_wav / p_wav.sum()
+
+        pulse.birth_times = sc.array(
+            dims=['event'],
+            values=np.random.choice(x_time, size=pulse.neutrons, p=p_time),
+            unit='s',
+        )
+        pulse.wavelengths = sc.array(
+            dims=['event'],
+            values=np.random.choice(x_wav, size=pulse.neutrons, p=p_wav),
+            unit='angstrom',
```
Duplicate code... or not?

nvaytet replied:
Yes, see update.

SimonHeybrock commented on src/tof/result.py:
```diff
@@ -0,0 +1,315 @@
+        self._arrival_times = MappingProxyType(self._arrival_times)
+
+    @property
+    def choppers(self) -> MappingProxyType[str, Chopper]:
+        """The choppers in the model."""
+        return self._choppers
+
+    @property
+    def detectors(self) -> MappingProxyType[str, Detector]:
```
Should this be the readonly variants now?

SimonHeybrock commented on src/tof/result.py:
```diff
@@ -0,0 +1,315 @@
+        return self._detectors
+
+    @property
+    def pulse(self) -> Pulse:
```
```suggestion
    def pulse(self) -> ReadonlyPulse:
```
?

SimonHeybrock commented on src/tof/result.py:
```diff
@@ -0,0 +1,315 @@
+    def __iter__(self):
+        return chain(self._choppers, self._detectors)
+
+    def __getitem__(self, name) -> Union[Chopper, Detector]:
```
readonly?

nvaytet commented:
@SimonHeybrock I think I am now more or less happy with this.
I think I addressed your final comments.

In addition, I think I fixed something that had been bugging me for some time.
Sampling the neutrons from the time and wavelength distributions were being done with `np.random.choice`, which only selects from the available values you pass to it, meaning that if you give 10,000 possible wavelengths, it will only have those in the pulse and nothing in between those values.

This meant that if you use a fine binning to inspect your result, you will see spikes separates by empty bins (Premek actually saw this in notebook) unless you really crank up the `sampling` number for the interpolation.

I tried out the [Scipy methods](https://docs.scipy.org/doc/scipy/tutorial/stats/sampling.html) for sampling from a continuous distribution (you just provide a callable function that defines the probability distribution instead of a discrete numpy array), but they are too slow for our use case (many minutes to create 10M neutrons).

Finally, I came up with a small trick that after sampling using `random.choice`, I add some noise to the values (+/- a small delta of the scale of the bin width) to spread them between the bins.
This seems to work well, but maybe you want to check there isn't something fundamentally bad with the approach?

SimonHeybrock commented on src/tof/pulse.py:
```diff
@@ -13,18 +13,198 @@
+
+        t = np.random.choice(
+            p_time.coords[t_dim].values, size=size, p=p_time.values
+        ) + np.random.uniform(-dt, dt, size=size)
```
Have you considered using a normal distribution?

nvaytet replied:
You mean for spreading the values? i.e. instead of the `random.uniform`?
I thought of it but decided it would make bumps instead of spreading evenly, and if the sampling is low (e.g. if you end up selecting only a small wavelength range using choppers), you would see them? (a bit like the spikes and the empty bins, but granted to a lesser degree)

SimonHeybrock replied:
My thought was concern potentially more artifacts with uniform (and overlapping) boxes, and generally better properties you get with gaussians, since it is basically you you do when smearing, and it approximates many actual physical effects.

nvaytet replied:
So the boxes here should not overlap because their ranges should meet exactly in the middle between two values.
But I will give it a go and see what it gives.

SimonHeybrock replied:
So you always have constant bin size in the distribution? Is that checked anywhere? Does the `numpy.choice` make that assumption?
